ecommerce/serializers.py

```diff
@@ -367,15 +367,29 @@ def _update_items(cls, basket, items):
             item = items[0]
             product_id = item.get("product_id")
             run_ids = item.get("run_ids")
-            product = models.Product.objects.get(id=product_id)
+
+            readable_id = None
+            try:
+                product_id = int(product_id)
+            except ValueError:
+                readable_id = product_id.replace(" ", "+")
+
+            if readable_id:
+                product = models.Product.objects.filter(
+                    dj_models.Q(courseruns__courseware_id=readable_id)
+                    | dj_models.Q(programs__readable_id=readable_id)
+                ).first()
+            elif product_id:
+                product = models.Product.objects.get(id=product_id)
```

Comment:
I agree with Asad's previous comment, and I also think this whole block should be an API/util function so we're not repeating it:

```python
def get_product_from_querystring_id(qs_product_id):
  if qs_product_id.isdigit():
    return models.Product.objects.get(id=int(qs_product_id))
  else:
    # Text IDs for Programs/CourseRuns have '+' characters, which represent spaces in URL-encoded strings
    product_text_id = qs_product_id.replace(" ", "+")
    return models.Product.objects.get(
      dj_models.Q(courseruns__courseware_id=product_text_id)
      | dj_models.Q(programs__readable_id=product_text_id)
    )
```

Reply:
@gsidebo it can be added in util function but the query is quite different in both of the cases. 

In `validate_items` function we are using `all_objects` product model manager like `models.Product.all_objects.filter()` which pick all of the records including inactive ones.

But in `_update_items` function we are using `objects` product model manager like `models.Product.objects.filter()` which only picks active records.

That's the reason we prefer to keep it here rather than putting in utils.

If you still suggest, I can add it.

Reply:
By the way, I'm changing it as you requested.